### PR TITLE
Functions: decode_tfvars, encode_tfvars, encode_expr on bult-in provider for compatibility

### LINKED ISSUE #1614: "terraform" provider functions: encode_tfvars, decode_tfvars, encode_expr
### OpenTofu Version

```shell
GitHub repo on main branch
```

### Use Cases

This is to follow up the discussions on Slack, https://opentofucommunity.slack.com/archives/C05RKE7PS72/p1714960161198039. Janos would like to use an issue for migration guides. There are 3 new functions introduced in Terraform 1.8:
```
"encode_tfvars"
"decode_tfvars"
"encode_expr"
```
More details: https://developer.hashicorp.com/terraform/language/functions/terraform-encode_tfvars

### Attempted Solutions

N/A

### Proposal

_No response_

### References

_No response_

### PR DESCRIPTION
**Pull request by Gogotchuri**
Implementations for built-in provider functions:
- decode_tfvars
- encode_tfvars
- encode_expr

I've included the last one because the implementation was quite straightforward, if we decide against including it, that's also fine.

P.S. I haven't added documentation to the website. I'm not certain if we should add it during this PR or create another PR later on.

Resolves #1614 

## Target Release

1.10.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

- [x] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

### CODE REVIEW

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:112]:
```suggestion
			Description:       "Input to encode for TFVars file. Must be an object with key that are valid identifiers",
```

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:119]:
```suggestion
		Summary:           "Encode an object into a string with the same format as a TFVars file",
```

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:145]: Is this TODO relevant?

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:169]:
```suggestion
			Description:       "Expression to encode",
```

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:78]: Does that function support non vars file formats? We might want to additionally check if that's the case and error for non-supported files

**Review by ollevche — COMMENTED** [review]
Looks good, thanks! Left a few questions

**Comment by ollevche** [comment]
Regarding the docs - it is fine to add them to this PR. We don't have a strict policy, but it is better to keep everything in one place, if that's not a huge amount of content to review.

File: internal/builtin/providers/tf/provider_functions.go
Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go:78]: Not anymore no, thanks for catching that!

File: internal/builtin/providers/tf/provider_functions.go
Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go]: Not anymore no, thanks for catching that!

File: internal/builtin/providers/tf/provider_functions.go
Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go]: Good point, more idiomatic to declare where it is actually needed, moved!

File: internal/builtin/providers/tf/provider_functions.go
Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go:57]: Do you refer to the file extension? 
They pass in the file's content, not the actual file. ParseConfig ensures the content is in the correct HCL syntax. 
Did I understand your question correctly?

**Comment by Gogotchuri** [comment]
Alright, I'll get started on Docs

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [reply internal/builtin/providers/tf/provider_functions.go:57]: Yep, I thought the `hclsyntax.ParseConfig` could parse something more generic than just a var file syntax. I double checked other usages and it seems fine to me.

**Comment by ollevche** [comment]
IMO the first option is a better choice. I would also put a short paragraph about this special case in the index of functions folder: https://opentofu.org/docs/language/functions/

**Review by ollevche — COMMENTED** [review]
Nice idea to create a separate page for the built-in provider! Left a few suggestions for documentation, overall looks good to me

File: website/docs/language/providers/builtin.mdx
Comment (Gogotchuri) [reply website/docs/language/providers/builtin.mdx:13]: Not mine, it was there, but thought the same, we should fix it probably

File: website/data/language-nav-data.json
Comment (Gogotchuri) [reply website/data/language-nav-data.json:153]: Neither do I, I will try to see what's up

File: website/docs/language/functions/index.mdx
Comment (Gogotchuri) [reply website/docs/language/functions/index.mdx:67]: Removing the usage example

File: website/data/language-nav-data.json
Comment (Gogotchuri) [reply website/data/language-nav-data.json:151]: Didn't work

Comment (Gogotchuri) [reply website/data/language-nav-data.json:151]: We should probably figure out navigation ordering issues in a separate issue

Comment (ollevche) [reply website/data/language-nav-data.json:151]: Docusaurus is often weird in its behaviour. I tried to put a sidebar_position in each page (Provider Configuration, Provider Requirements and Built-in Provider) and it worked!

Comment (Gogotchuri) [reply website/data/language-nav-data.json:151]: Done, it worked!

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:64]: We are silencing warnings everywhere, is it intended?

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [inline internal/builtin/providers/tf/provider_functions.go:161]: Not sure if that's enough from UX perspective. What do we usually output if the value is not wholly known?

Also, makes sense to double check other error messages (however, others look good to me).

Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go:161]: In a few places, we output that it is not wholly known, in other cases, we say that the value isn't known until the apply.
Do you have any suggestions on what we should output in this case?

Comment (ollevche) [reply internal/builtin/providers/tf/provider_functions.go:161]: I think this is completely fine then to have an error message saying "input is not wholly known", it would result in something like:

```
│ Error: Error in function call
│ 
│   on main.tf line 10, in output "name":
│   10:   value = provider::terraform::encode_tfvars(var.something)
│     ├────────────────
│     │ while calling provider::terraform::encode_tfvars(input)
│ 
│ Call to function "provider::terraform::encode_tfvars" failed: input is not wholy known.
```

File: internal/builtin/providers/tf/provider_functions.go
Comment (ollevche) [reply internal/builtin/providers/tf/provider_functions.go:64]: It seems safe to do so since the protocol itself does only allow errors to be thrown. Also, we shouldn't expect any warnings here AFAIC.

**Review by ollevche — DISMISSED** [review]
Looks good to me, thanks!

I would like to have a second review before merging this one.

File: internal/builtin/providers/tf/provider_function_test.go
Comment (Evi1Pumpkin) [inline internal/builtin/providers/tf/provider_function_test.go:1]: I believe the name of the test file should be `provider_functions_tests.go` (missing an `s` in "functions")

Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_function_test.go]: @Evi1Pumpkin Changed, thanks for catching that!

File: internal/builtin/providers/tf/provider_functions.go
Comment (Evi1Pumpkin) [inline internal/builtin/providers/tf/provider_functions.go:92]: This is a nitpick, but I was wondering if we can put a more specific type here in the input type, as we know that the input should be an `object`

File: internal/builtin/providers/tf/provider_functions.go
Comment (Evi1Pumpkin) [reply internal/builtin/providers/tf/provider_functions.go:150]: The word `expression` is still lowercase

File: internal/builtin/providers/tf/provider_functions.go
Comment (Gogotchuri) [reply internal/builtin/providers/tf/provider_functions.go:92]: I went through the docs for cty types. There doesn't seem to be any other way to have an object type without knowing it's concrete type at compilation time.
If you can find more descriptive type for our usecase, that would definitely be better.

Comment (Evi1Pumpkin) [reply internal/builtin/providers/tf/provider_functions.go:92]: I looked at this a little bit but didn't found anything, so let's drop this

**Review by Evi1Pumpkin — APPROVED** [review]
Great job! Thank you. Great solutions!

**Pull request merged**
